pkg/cloud/gcp/actuators/machine/actuator.go

```diff
@@ -96,7 +112,21 @@ func (a *Actuator) Exists(ctx context.Context, machine *machinev1.Machine) (bool
 	// When create()/update() try to store machineSpec/status this might result in
 	// "Operation cannot be fulfilled; the object has been modified; please apply your changes to the latest version and try again."
 	// Therefore we don't close the scope here and we only store spec/status atomically either in create()/update()"
-	return newReconciler(scope).exists()
+	_, exists, err := newReconciler(scope).exists()
+	if isInvalidMachineConfigurationError(err) {
+		// If the machine has, e.g. invalid zone, and it doesn't have a phase set yet,
+		// we have to make sure the phase goes as "Failed".
+		if machine.Status.Phase == nil {
+			machine.Status.Phase = pointer.String("Failed")
+		}
+		// If the machine has, e.g. invalid zone, and we delete the invalid machinset,
+		// we want to set the machine to "Deleting" phase and return nil as error.
+		// We need the error to be nil so we can successfully delete the invalid machine.
+		if *machine.Status.Phase == "Deleting" {
+			return false, nil
+		}
```

Comment:
Hmm, this worries me, what if there is actually a Machine in the cluster, then we update the configuration to be invalid, say, the credentials expire, now we remove the Machine object, but the Machine itself remains in the cloud provider right?

If there's any chance of leaking machines we should probably close this as a won't fix

Reply:
So are you talking about a machine being in phase `Running` and then we edit it, so it has an invalid configuration?
I have not tried that. Will investigate. Nevertheless, what you are saying makes sense.

Reply:
What we could theoretically do is: In function `exists()`, we would try and get that machine, if that failed, we would know that the machine was never created properly in the first place and then work with that information if it exists. What do you think?

Reply:
> So are you talking about a machine being in phase Running and then we edit it, so it has an invalid configuration?
I have not tried that. Will investigate. Nevertheless, what you are saying makes sense.

Yes, that's one of my concerns, if the VM exists, ie we have a provider ID set, then we must make sure, even if the configuration is no longer valid, that we don't remove the machine until we are certain the VM is gone

> we would try and get that machine, if that failed, we would know that the machine was never created properly in the first place and then work with that information if it exists. What do you think?

I think we can rely on the providerID being set, there are many reasons why a call to get the instance might fail, for example, if the user mistakenly changed the region or zone, we know the information in the Machine isn't right, but because there is a providerID, we know we did create a VM. 

By not removing the Machine in this case, we alert the user that some manual intervention is required, which I think is desired

Reply:
Yes, the only thing that I had to do different is the part about `providerID`. Even machines that are not created properly have some. That's why I opted for node reference instead.

Reply:
>  Even machines that are not created properly have some

Can you expand on this? A providerID should only ever be set once GCP has accepted the instances create request, ie, we've told it to create a VM, it has started that process

Reply:
I didn't go deep into this, but I would expect this expression to be true `providerID == ""`, if the machine is not created properly, but that is not the case. So I just looked into differences in config of valid and invalid machines and came up with the node ref solution.

Reply:
How did you simulate the machine not being created properly?

Reply:
1. `oc get machineset <ms-name> -o yaml > invalid.yml`
2. Change zone, name and labels from `us-central-1a` to `us-central-1d`
3. `oc create -f invalid.yml`

Reply:
I just tried this with the current code from master, I get the following log messages
```
I1214 15:13:03.073223       1 controller.go:187] jspeed-test-kq9kx-worker-f-tcxst: reconciling Machine
I1214 15:13:03.073259       1 actuator.go:84] jspeed-test-kq9kx-worker-f-tcxst: Checking if machine exists
E1214 15:13:03.205859       1 controller.go:297] jspeed-test-kq9kx-worker-f-tcxst: failed to check if machine exists: jspeed-test-kq9kx-worker-f-tcxst: Machine does not exist
E1214 15:13:03.213819       1 controller.go:326]  "msg"="Reconciler error" "error"="jspeed-test-kq9kx-worker-f-tcxst: Machine does not exist" "controller"="machine-controller" "name"="jspeed-test-kq9kx-worker-f-tcxst" "namespace"="openshift-machine-api" "object"={"name":"jspeed-test-kq9kx-worker-f-tcxst","namespace":"openshift-machine-api"} "reconcileID"="5414d86b-1a74-4c76-9567-f06b1b1067b3"
```

The providerID is not set on the machine and the phase remains empty. So the providerID should be ok for invalid machines as far as I can tell